### PR TITLE
refactor: SDK workflow builds

### PR DESCRIPTION
**Pull request by CodeSchwert**
### Hi👋, please prefix this PR's title with:

- [ ] `breaking-change:` if you have introduced modification that necessitates immediate adjustments by this SDK's users to their applications, clients, or integrations to avert disruptions to existing features or functionalities.
- [x] `feat:`, `fix:`, `refactor:`, `docs:`, or `chore:`.

# Summary

Refactor the Build, Lint, Typecheck, Test workflow re-use build and `node_module` artefacts across steps rather than re-running the install and build steps on every job.

# Detail and impact of the change
- No impact to external customers

## Changed
- Cache build and `node_modules` in Build, Lint, Typecheck, Test workflow

# Anything else worth calling out?
- More caching to follow up after this PR is merged

### CODE REVIEW

**Review by wdinau — COMMENTED** [review]
Hi Nik, just some newbie questions
Q1: is this the only place to use cache (compare to all other places in .github/workflows) ?
Q2: we put cache node modules on the top in `build`, and all subsequent steps are using that cache.would there ever be cases where people re-run part of the job, (say `build-passport`) without running the `build`, and they'd get to use the latest build of their change?  
Q3: ` build-cache-sdk` (in build-sdk.yaml) is that a step `build` job (of build-lint-typecheck-test.yaml) can re-use instead of copy paste same logic again?

**Comment by CodeSchwert** [comment]
> Hi Nik, just some newbie questions Q1: is this the only place to use cache (compare to all other places in .github/workflows) ? Q2: we put cache node modules on the top in `build`, and all subsequent steps are using that cache.would there ever be cases where people re-run part of the job, (say `build-passport`) without running the `build`, and they'd get to use the latest build of their change? Q3: ` build-cache-sdk` (in build-sdk.yaml) is that a step `build` job (of build-lint-typecheck-test.yaml) can re-use instead of copy paste same logic again?

Q1: yes, currently only the build-lint-typecheck workflow is using the cache. We can expand to more workflows, but makes the most sense to update this one first since it has multiple `build` steps in the same workflow.
Q2: The node_modules cache is based off the yarn.lock changing, which is unlikely to change during the workflow. Infact, the workflow is designed to error out if the `yarn.lock` file changes.
Q3: I'm planning to use the `build-sdk` workflow to share the build artefacts across workflows. Apparently it needs to be in the main branch first before I can do that... so including it in this PR for future use. We can potentially reuse both `node_modules` and `build` artefacts across all the workflow runs in the PR.

File: .github/workflows/build-sdk.yaml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+          cache: 'yarn'
+
+      - name: Install dependencies
+        run: yarn install --immutable
+
+      - name: Cache build artifacts
+        uses: actions/cache@v4
+        with:
+          path: node_modules
+          key: ${{ runner.os }}-build-cache-deps-${{ hashFiles('yarn.lock') }}
```
Comment (wdinau) [inline .github/workflows/build-sdk.yaml:30]: where is `build-cache-deps` coming from?
I don't know enough but i can see a pattern here the key is build-cache-xxx and the job is called `build-cache-sdk` is there another key/job exist with the word `deps` ?

Comment (CodeSchwert) [reply .github/workflows/build-sdk.yaml:30]: Not yet, that will be a follow up PR to use that cache once this workflow is merged to main.

**Review by wdinau — APPROVED** [review]

**Pull request merged**